### ISSUE TITLE
log/slog: make the current context easily available to loggers 

### ISSUE THREAD

**Issue by jba** [golang/go#58243]
This issue assumes that the [slog proposal](https://go.dev/issue/56345) will be accepted.

# Background

Adding dynamically scoped information to log output greatly aids debugging: it allows logs to be grouped by server request, and gives readers valuable clues about what was happening along the control flow path that led to the log event. One important piece of information is the _trace ID_, which is added to the context by tracing systems like OpenTelemetry to construct tracing graphs, a valuable performance and debugging tool. If a log event contains a trace ID, it can be used to annotate the trace in the graph, making it easier to see what was going on in the trace. We’ll focus on the trace ID in this issue for concreteness, but whatever solution emerges from this discussion should work for any value in a `context.Context`.

In OpenTelemetry, trace IDs are contained in _spans_. A trace is started with a call like
```
ctx, span := trace.Start(ctx, name, ...)
```
That call starts a new span, makes it a child of the currently active span (if any), and returns a context containing the span as well as the span itself. It is typically followed by
```
defer span.End()
```
which ends the trace. In a server, a typical trace will be an entire request, but nested traces can be established anywhere along the call chain.

# The problem

The goal of this issue is to find a good way to convey the context to a slog.Logger, and ultimately to a slog.Handler, so the trace ID and other context values can appear in the output. We’ll assume the programmer has already arranged to pass the context itself around, so that all logging occurs inside a function with a context:
```
func f(ctx context.Context, ...) {
```
How can we encourage the programmer to include the context in their logging statements?

The key word here is _encourage_. The `slog` package already has a way to include the context:
```
logger.WithContext(ctx).Info(msg)
```
but programmers are likely to prefer the shorter
```
logger.Info(msg)
```

We also need to consider the related problem of how the logger itself is passed around. Although `slog` has a global logger, that logger will not have attributes that may have been added along the call chain with lines like
```
logger = logger.With("key", "value")
```
So programs that care about dynamically scoped data will have to pass the logger down the stack along with the context.

# Loggers in contexts

`slog`'s original solution to this problem was to include the logger itself in the context. `NewContext` would add the logger to the context:
```
ctx = slog.NewContext(ctx, logger)
``` 
`FromContext` would extract it:
```
logger := slog.FromContext(ctx)
```
Callers would still have to write
```
logger.WithContext(ctx).Info(msg)
```
but we also included a convenience function, `Ctx`, that combined `FromContext` and `WithContext`:
```
slog.Ctx(ctx).Info(msg)
```
Before we reject this solution and look elsewhere, let’s see what’s good about it. It simplifies the code, because instead of passing around both a context and logger, only the context needs to be passed. And it encourages programmers to include the context in their log output, because `slog.Ctx(ctx)` is a more compact way to obtain a logger than `slog.FromContext(ctx)`.

The author of a tracing system could make it even more likely that log output would contain trace IDs, by modifying the `Start` method that adds a span to a context. That code could also add the trace ID to the logger in the context:
```
func Start(ctx context.Context, ...) (context.Context, Span) {
    ...
    logger := slog.FromContext(ctx)
    logger = logger.With("traceID", traceID)
    return slog.NewContext(ctx, logger), span
}
```
Now the trace information would be logged as long as the programmer retrieved the logger from the context.

# Controversy

Some comments on the proposal issue were in favor of including a logger in a context: [@v3n](https://github.com/golang/go/issues/56345#issuecomment-1289940433), [@rliebz](https://github.com/golang/go/issues/56345#issuecomment-1322244859), [@darkfeline](https://github.com/golang/go/issues/56345#issuecomment-1367640591), [@neild](https://github.com/golang/go/issues/56345#issuecomment-1367739614) [@dottedmag](https://github.com/golang/go/issues/56345#issuecomment-1379553502), [@tigrannajaryan 1](https://github.com/golang/go/issues/56345#issuecomment-1398799820), [2](https://github.com/golang/go/issues/56345#issuecomment-1398982729), [3](https://github.com/golang/go/issues/56345#issuecomment-1405873193), [@Aneurysm9](https://github.com/golang/go/issues/56345#issuecomment-1399075328).

But others were strongly against it: [@prochac 1](https://github.com/golang/go/issues/56345#issuecomment-1289572644), [2](https://github.com/golang/go/issues/56345#issuecomment-1336545767), [3](https://github.com/golang/go/issues/56345#issuecomment-1292752146), [​​@mminklet 1](https://github.com/golang/go/issues/56345#issuecomment-1289842738), [2](https://github.com/golang/go/issues/56345#issuecomment-1292756841), [3](https://github.com/golang/go/issues/56345#issuecomment-1360861788), [@rabbit](https://github.com/golang/go/issues/56345#issuecomment-1292672956), [@jellevandenhooff](https://github.com/golang/go/issues/56345#issuecomment-1352286205), [@ChrisHines](https://github.com/golang/go/issues/56345#issuecomment-1379438240).

The main objection is that putting a logger in a context creates a hidden dependency. When you call a function that takes a context, you don’t know whether the context is for logging, cancellation, or both. Context arguments are added to functions just to allow logging. The objectors understand that adding loggers to contexts might make sense for some applications, but oppose adding it to the standard library for fear of creating a precedent that will lead to widespread abuse of contexts.

Because of these concerns, we removed `NewContext`, `FromContext` and `Ctx` from the proposal. So to obtain a logger that may contain dynamically scoped information, a function must take it as an argument somehow, either directly or as a part of another argument, like a receiver field.

# Alternatives

If we don’t add loggers to contexts, then how can we encourage programmers to pass contexts to loggers?

We could do nothing to the API, but educate users to write
```
logger = logger.WithContext(ctx)
```
at the top of their functions. This might not seem much different than remembering to write `logger := slog.FromContext(ctx)`, but note that previously, `FromContext` was the only reasonable way to get a logger with dynamic content. Now we assume a logger is already available to the function, so nothing pushes the programmer towards adding the context to it.

Another popular idea is to require each log output method to take a context:
```
func (*Logger) Info(ctx context.Context, message string, ...)
...
```
We don’t want to do that, because it goes too far in the other direction, forcing programmers to pass in a context (or at least consider doing so) for every log message.

A related idea is to add context-aware functions to the ones already present:
```
func (*Logger) Info(message string, ...) // already in the proposal
func (*Logger) InfoCtx(ctx context.Context, message string, ...)
...
```
We also don’t want to do that, because it greatly increases the size of the API. It’s also not much different from the existing `WithContext` method, especially if we shortened the name to `Ctx`:
```
logger.InfoCtx(ctx, "message")
logger.Ctx(ctx).Info("message")
```

Another variant of this idea is an adapter type, like @[jellevandenhooff](https://github.com/jellevandenhooff)’s [slogctx package](https://github.com/golang/go/issues/56345#issuecomment-1407785642). You would then write
```
clogger := slogctx.NewLogger(logger)
...
clogger.Info(ctx, "message")
```

All of these solutions require work in the function doing the logging. What if we could move that work up the call stack? The caller could add the context to the logger whenever the context changed:
```
// Add a span to the context.
ctx, span := tracer.Start(ctx, ...)
defer span.End()
// Add the context to a logger.
logger = logger.WithContext(ctx)
// Pass the logger down.
f(ctx, logger)
```
Now `f` can use the logger directly and still get trace IDs in the output. Since context changes occur relatively infrequently, this convention should reduce the total amount of programmer toil. More importantly, it puts the responsibility in the hands of the programmer who understands the importance of tracing, and how trace IDs are propagated through contexts.

~~But we can do even better by using an idea first suggested on the proposal issue by [@rabbit](https://github.com/golang/go/issues/56345#issuecomment-1292672956). If we are willing to mutate a part of the context, we can avoid having to repeatedly incorporate the context into the logger. I sketched an [implementation](https://github.com/jba/slog/blob/main/trace/trace.go) for OpenTelemetry that associates a list of spans with a context key. Starting a new span appends it to the list, and ending a span removes it from the list. Once a logger contains a context with the list, it can be used repeatedly. That means that `logger.WithContext` needs to be called only for the top-level span. Starting a nested span can be done without updating the logger, since doing so mutates (a part of) the context that the logger already contains.~~

~~There is one difference in behavior between the purely functional implementation of adding spans to contexts and the mutating one: if a context is discarded (as by function return) without calling `span.End`, the functional implementation will drop the span while the mutating one will not. I’m not sure if that’s a bug or a feature.~~

[Edit: mutating the list of spans in the context does the wrong thing if spans don't follow a stack discipline. It is not recommended for general use.]

Do any of these alternatives adequately ensure that most log events contain trace information? Are there other ideas I’ve missed? Is there anything we can add to or change about `slog` to help?

**Comment by stephenafamo** [comment]
While this can work, I personally don't like it because it makes the logger a "request scoped dependency" when it is only a few values in the context that should be.

For example, when writing a web service, I will often have a struct that look like

```go
type Server struct {
    DB     store.DB
    Logger log.Logger // whichever log package I'm using
    // Other dependencies
}

func (s Server) Users(w http.ResponseWriter, r *http.Request) error {
    // .....
    // If I need to log, I will use it like this
    s.Logger.Log(...)
}
```

With your suggestion, a logger will have to be created at the top of the request tree and passed down.

This means that aside from the context, every function that needs to log also needs to accept a logger since setting it as a one-time dependency is no longer possible.

Personally, I think this is even more cumbersome than doing `s.Logger.WithContext(ctx).Log(...)`.

I much prefer the ergonomics of using the `slogctx` wrapper and will probably use that (or something else) in cases where I need the context for logging. Which is why I think it will be useful to include it as part of the standard package (or a sub package).

**Comment by icholy** [comment]
What if, instead of passing the whole logger in the context, we only pass the attributes? Then the handler can include them the same way it includes the span info.

**Comment by tigrannajaryan** [comment]
> How can we encourage the programmer to include the context in their logging statements?
> The key word here is _encourage_.

Thanks for opening this issue @jba
 
If I understand your proposal correctly you suggest that the `Logger` is passed as a parameter to every function that needs to log. Can you please clarify how we will _encourage_ the programmer to accept the `Logger` as a parameter? Is documentation going to recommend that the `Logger` is never stored in a struct and is always passed as a parameter, like we [already do](https://pkg.go.dev/context) for `Context`?

**Comment by komuw** [comment]
`slog.Logger` has three `With*` methods; `WithGroup`, `With`, `WithContext`.     
Of the three, two of them(`WithGroup`, & `With`) call underlying methods on the handler: https://github.com/golang/exp/blob/98cc5a0785f9389a91393b250df90a27725107fc/slog/logger.go#L103-L115    
`WithContext` is the odd one out, it doesn't call any method on the handler.    

In order to aid in propagating things like `traceID/spanID`, I think it would be worthwhile for `WithContext` to call some underlying handler method.   
Assume we expanded the handler interface
```go
type Handler interface {
        .... 
	WithContext(ctx context.Context) context.Context
}
```
and then `logger.WithContext` becomes;
```go
func (l *Logger) WithContext(ctx context.Context) *Logger {
	c := l.clone()
	c.ctx = l.handler.WithContext(ctx)
	return c
}
```

This would enable people to write custom handlers that do things like;
```go
type customHandler struct{ traceID string}
func (c customHandler) WithContext(ctx context.Context) context.Context {
    if vCtx := ctx.Value(myTraceIdCtxKey); vCtx == nil {
       // we got an un-annotated context, let's add traceID to it
       ctx = context.WithValue(ctx, myTraceIdCtxKey, c.traceID)
   }

  return ctx
}
```

**Comment by jba** [comment]
@stephenafamo, @icholy: one advantage of passing the logger around is the ability to use
```
logger = logger.With(...)
```
which may preformat some attributes (depending on the handler; the built-in ones do). That can result in a big speedup.

**Comment by stephenafamo** [comment]
@jba I don't know the internals of slog, so I need to clarify:

In a hypothetical HTTP server, are you saying:

* Create a logger for each request and pass it down
**May be more performant than**
* Have a single instance of the logger as a dependency

**Comment by jba** [comment]
> If I understand your proposal correctly you suggest that the Logger is passed as a parameter to every function that needs to log. Can you please clarify how we will encourage the programmer to accept the Logger as a parameter? Is documentation going to recommend that the Logger is never stored in a struct and is always passed as a parameter, like we [already do](https://pkg.go.dev/context) for Context?

@tigrannajaryan, I don't have a proposal here. Personally, I think it's fine to put a logger in a context, but we've heard enough serious, thoughtful disagreement about it to start a new issue to consider alternatives. I am hoping to hear from some of the people who don't like loggers in contexts, like @prochac, @mminklet, and others, how they provide access to loggers. @ChrisHines, you wrote "We felt the practice of passing loggers in contexts was harmful enough that we are now undergoing an effort to eliminate it from our code base." What are you doing instead?

**Comment by jba** [comment]
@seankhliao, you're right about both of those points. The mutating implementation is not a good idea.

**Comment by AndrewHarrisSPU** [comment]
> @seankhliao, you're right about both of those points. The mutating implementation is not a good idea.

I found the direction suggested by this mutating implementation appealing, at least. A more narrowly tailored API for managing data that should be logged and should live on a context (as opposed to data that shouldn't be logged but should live on a context, or data that should be logged but shouldn't live on a context) seems useful.

Eventually I wonder if this is related to the third point of https://github.com/golang/go/issues/56345#issuecomment-1413742844, about when a `LogValuer` is evaluated. Would it make sense to similarly define when, for a given `Record`, all observations drawn from a context are evaluated? Along the lines of: invariantly, the first `Handler` a `Record` passes through after being created should traverse the `Record`, evaluate and add attributes based on observing the context. The invariant could be folded into a `Handler` that just isolates the context-related evaluation, or it could be folded into a `Handler` that does more.

This would be different from other `Handler`s because it would need to maintain a list of keys to look up. This kind of a handler would have additional behavior and state, like having a `WithContextKey` in addition to `WithGroup` and `WithAttrs` (although a `WithContextKey` wouldn't be eagerly rendering formatted/encoded text.) But, because this is a proper subtype relationship and a `Handler` is an interface, we could pass it as a vanilla `Handler` - it would just be no-ops when there is no context or no keys. This makes sense if only the context given at the time of creating a `Record` matters and I'm not sure that's true, I just don't see why it wouldn't be.

**Comment by jba** [comment]
@komuw, we want WithContext to be very fast, so we don't want to clone anything. We want people to be able to write `logger.WithContext(ctx).Info(msg)`.

**Comment by jba** [comment]
> In a hypothetical HTTP server, are you saying:
> - Create a logger for each request and pass it down
> May be more performant than
> - Have a single instance of the logger as a dependency

@stephenafamo: Yes, if there are attributes that are common along a call chain.

If you have a fixed logger, you have to add those common attributes in each log call:
```
func (s *Server) handle(...) {
    commonAttrs := []slog.Attr{...}
    s.part1(commonAttrs)
    s.part2(commonAttrs)
}

func (s *Server) part1(as []slog.Attr) { s.logger.Info(msg, slog.String("func", "part1"), as...) }
func (s *Server) part1(as []slog.Attr) { s.logger.Info(msg, slog.String("func", "part2"), as...) }
```

The common attrs will be formatted on each call.

If instead you wrote:
```
func (s *Server) handle(...) {
    logger := s.logger.With(commonAttrs...)
    s.part1(logger)
    s.part2(logger)
}

func (s *Server) part1(l *slog.Logger) { l.Info(msg, slog.String("func", "part1")) }
func (s *Server) part1(l *slog.Logger) { l.Info(msg, slog.String("func", "part2")) }
```
then the common attrs would be formatted only once, in the `logger.With` call. (That is true for the built-in handlers, and for any Handler that does a good job of implementing `WithAttrs`.)

**Comment by komuw** [comment]
> @komuw, we want WithContext to be very fast, so we don't want to clone anything. We want people to be able to write `logger.WithContext(ctx).Info(msg)`.

@jba `WithContext` already clones: https://github.com/golang/exp/blob/98cc5a0785f9389a91393b250df90a27725107fc/slog/logger.go#L123-L124, unless you meant something else.

**Comment by jba** [comment]
@komuw, you're right, it clones the Logger. I was thinking of cloning the Handler. That is typically expensive, but as your example shows, it may not be necessary.

However, I don't really understand your example. Can you give a practical application of Handler.WithContext? Maybe you've given one, but I don't see how the traceID field on your handler gets set.

**Comment by tenntenn** [comment]
>This issue assumes that the [slog proposal](https://github.com/golang/go/issues/58243#56345) will be accepted.

Is this URL correct?

**Comment by komuw** [comment]
> @komuw, 
> However, I don't really understand your example. Can you give a practical application of Handler.WithContext? Maybe you've given one, but I don't see how the traceID field on your handler gets set.

@jba I took sometime to try and sketch out an implementation and it didn't work out as I had hoped. I take back my suggestion.

**Comment by komuw** [comment]
Thinking about this issue a little more.   
It asks two questions;

> 1. How can we encourage the programmer to include the context in their logging statements?

I think the answer to this question is easy if you paraphrase the question as follows; How can we encourage the programmer to include the context in their function calls?     
The `net/http` package has two general ways to create a request; ` NewRequest()` & `NewRequestWithContext()`. How do we normally encourage people to use the latter over the former? 
The answer to that is also the answer to the previous question about logging and contexts.

> 2. How is the logger itself passed around? Although slog has a global logger ....

The answer to this is; either remove the global logger or document that whoever chooses to use the global logger is opting into that behaviour.

All in all, I feel like the solutions to these questions are; documentation, custom style guides & custom linters.

**Comment by sagikazarmark** [comment]
Personally, I like `logger := s.logger.WithContext(ctx)` where `s` is the receiver of a method on a service struct. I generally inject loggers as dependencies to my services and call `WithContext` whenever I need it.

I wholeheartedly agree with the "hidden dependency" argument.

While I think having a single, idiomatic way to solve a problem is better, having an alternate wrapper around the logger that exposes `logger.Info(ctx, ...)` methods as a syntactic sugar calling `logger.WithContext` under the hood is fine IMO.

Lastly, I think the question that needs answering is who is the intended target of any of the solutions being discussed here. Most applications that I consider well-architected have some sort of separate service layer from the transport (HTTP, gRPC, etc) and the first parameter passed to those services is always the context. People who has the discipline to build applications like that will probably use `logger.WithContext` if it's available and will probably prefer a single solution.

Since a context may not always be present (eg. in `main`), I'd prefer always writing `logger.Info(...)` compared to switching between that and `logger.Info(ctx, ...)` or `logger.InfoCtx(ctx, ...)`.

That leaves `logger.WithContext` as the winner in my eyes.

**Comment by jba** [comment]
@tenntenn, fixed, thanks.

**Comment by rabbbit** [comment]
Thanks for splitting this up and a whole write-up @jba! 

> The author of a tracing system could make it even more likely that log output would contain trace IDs, by modifying the Start method that adds a span to a context. That code could also add the trace ID to the logger in the context:
> 
> ```
> func Start(ctx context.Context, ...) (context.Context, Span) {
>     ...
>     logger := slog.FromContext(ctx)
>     logger = logger.With("traceID", traceID)
>     return slog.NewContext(ctx, logger), span
> }
> 
> ```
> Now the trace information would be logged as long as the programmer retrieved the logger from the context.

This could be perhaps the root of our "disagreement" - the code above would not work in the "style" I, and at least some parts of my corporation, tend to use loggers - loggers are persistent/long-lived/with extra-meta data. Thus to achieve ^ we'd need some kind of "Merge" function that takes two loggers - the long-lived & the per-request part. The same problem applies to `logger.WithContext` though.

I see how I'm biased/non-representative here, but if we were to assume that loggers are long-lived and contexts carry "...request-scoped values across API boundaries and between processes" then loggers would need to take contexts as any other argument. (I'm personally okay with simply enforcing ctx being present).

Some more random thoughts:
- we talk about spans/tracing here, but we have plenty of other metadata that we attach to per-request logging too. That data is sometimes, but not always, part of the tracing baggage.
-- in the context of the above, I don't right now fully see how `logger.WithContext` would know which values from context to log. This would mean some global registration of "known context keys", which seems messy.
- tangential: we had some internal linting against spurious use of `logger.With(...` for loggers that were then not logging anything. This turned out to be surprisingly costly for some applications.

**Comment by stephenafamo** [comment]
I have a suggestion. Let the output methods treat `context.Context` as a special case.

To do this, the slog package will define an internal context key whose values are a slice of `slog.Attr`. If one of the args given is a `context.Context`, the output method will retrieve the attached attributes on the context key and append them to the log message.

The slog package can then also include some helper methods to append attributes to this key.

```go
type ctxKey string
const ctxAttrs ctxkey = "attributes"

func AddToContext(ctx context.Context, attrs ...Attr) {
    existing, _ := ctx.Value(ctxAttrs).([]Attr)
    return context.WithValue(ctx, append(existing, attrs...))
}
```

Using it with open telemetry may then happen like this:

```go
func (s *server) aFunction(ctx context.Context) {
    // the otel package can include the traceID and spanID in the context
    // when creating the new span
    span, ctx := s.tracer.Start(ctx, "span_name") 
    
    // ...... 

    s.Logger.Info("somthing happened", ctx, slog.Int("status", 500))
}
```

As a bonus, I think this will also remove the need for the awkward context on log records since any values meant to be passed to that record would have been evaluated by the output method.  
`Logger.WithContext()` on the logger may also not be needed anymore since `Logger.With()` will work as expected with a context argument.

I believe this hits all the goals.

1. It does not require using a context *everywhere*
2. When we do want to use a context, it is very easy to do so since we don't have to resolve the keys and values manually
3. Things that modify the context, like middlewares or telemetry can add attributes to the context and pass it down. It will also be evaluated once when it is added to the context, not every time a downstream log message is emitted.

**Comment by jba** [comment]
@stephenafamo, nice idea. It does make it a little easier for people to put the context in their logs, but at the cost of adding another special case to key-value args. Also, what would we do for `LogAttrs`?

We could keep the second idea but not the first by just putting the `ctx` arg into `Record.Context`.

You can get the same effect in slog now by wrapping the context in a LogValuer, though it's more cumbersome to use:
```
type contextLogValuer struct {
	ctx context.Context
}

func (v contextLogValuer) LogValue() slog.Value {
	attrs, _ := v.ctx.Value(ctxAttrs).([]slog.Attr)
	return slog.GroupValue(attrs...)
}

func ContextAttrs(ctx context.Context) slog.LogValuer { return contextLogValuer{ctx} }

func f(ctx context.Context) {
    logger.Info("msg", "", ContextAttrs(ctx), ...)
}
```

**Comment by stephenafamo** [comment]
> @stephenafamo, nice idea. It does make it a little easier for people to put the context in their logs, but at the cost of adding another special case to key-value args. Also, what would we do for `LogAttrs`?

I think similar to `AddToContext`, the slog package should include an `ContextAttrs` function. 

```go
func ContextAttrs(ctx context.Context) slog.Attr {
	attrs, _ := v.ctx.Value(ctxAttrs).([]slog.Attr)
        // I'm not sure how slog treats empty group keys. I assume they are flattened
	return slog.Group("", attrs...) 
}

// Then it can be used with LogAttrs like this
slog.Info(msg, slog.ContextAttrs(ctx), slog.Bool("done", true))
```

> <span></span>
> We could keep the second idea but not the first by just putting the `ctx` arg into `Record.Context`.
>.

Records only have one context, what of if the output function is used with multiple contexts.  `slog.Info("", ctx1, ctx2, ctx3)`. 
Of course, this is not expected, but we would have to decide what happens in such a scenario.

Additionally, if the context is only used to pass request scoped log attributes, for what other reason will the record need to include the context?  
Perhaps treating the context as an "**Attr carrier**" allows us to remove the antipattern of including a context as part of the struct.


> You can get the same effect in slog now by wrapping the context in a LogValuer, though it's more cumbersome to use:
> 
> ```
> type contextLogValuer struct {
> 	ctx context.Context
> }
> 
> func (v contextLogValuer) LogValue() slog.Value {
> 	attrs, _ := v.ctx.Value(ctxAttrs).([]slog.Attr)
> 	return slog.GroupValue(attrs...)
> }
> 
> func ContextAttrs(ctx context.Context) slog.LogValuer { return contextLogValuer{ctx} }
> 
> func f(ctx context.Context) {
>     logger.Info("msg", "", ContextAttrs(ctx), ...)
> }
> ```

Exactly the problem. While the ingredients already exist, it is quite cumbersome to log values in the context.  
I believe this discussion is to find a way to make that easier.

**Comment by AndrewHarrisSPU** [comment]
@stephenafamo 
I like the idea of having one symbol in the context holding a segment or table of loggable values. To play nicely, middleware should prefer `AddToContext`, `ContextAttrs`, while end users should prefer middleware-provided API, is this the idea?

As far as multiple contexts appearing in a `Record` ... I feel like the all the smaller problems are significantly easier to reason about if there's at least a documented understanding of a 1:1 `Record`/`Context` approach, and anything trickier is off the map.

@jba
> If we don’t add loggers to contexts, then how can we encourage programmers to pass contexts to loggers?

Forgetting `WithContext` in a line like `log.WithContext(ctx).Info("...")` seems easy to do. Also, it might not be just "forgetting" ... sometimes it might be hard to reason locally about whether a logger is expecting to be handed a context, or whether it already has one that shouldn't be overwritten. A way to change nothing but suggest an optional style:

```go
func call(ctx context.Context, log func(context.Context) *slog.Logger){
    ...
    log(ctx).Info("...")
    ...
}

func main() {
    ...
    callWithContextAndLogger(ctx, log.WithContext)
    ...
}
```

The "change something" approach might be:

```go
type Pretext func(context.Context) *Logger

func (l *Logger) Pretext() Pretext {
    // mostly the same implementation as `WithContext`
}

func call(ctx context.Context, log slog.Pretext) { ... }

func main() {
    ...
    call(ctx, log.Pretext())
```

There are other definitions of `Pretext` (e.g., `type Pretext *slog.Logger`) that could work, the idea would be to mask the logger output methods until a later caller has provided a context.

**Comment by xorkevin** [comment]
Hi all! I'm relatively new to participating in forums like this, though I've been using go a long time and care a lot so thought to contribute to the discussion here.

@jba Thanks for being so open to feedback on this! Highly agreed with the concerns that we should not prefer to pass dependencies implicitly.

@stephenafamo, I like a lot of the ideas you're proposing here! I agree that instead of passing loggers implicitly in contexts, we should instead prefer something similar to what you wrote:

> ```go
> type Server struct {
>     Logger log.Logger // whichever log package I'm using
>     // Other dependencies
> }
> func (s Server) Users(w http.ResponseWriter, r *http.Request) error {
>     // .....
>     s.Logger.Log(...)
> }
> ```

There are a few places where I differ in opinion slightly though.

First is more of an implementation detail, but nevertheless something I think is worth calling out. With regards to the following:

> ```go
> func AddToContext(ctx context.Context, attrs ...Attr) {
>     existing, _ := ctx.Value(ctxAttrs).([]Attr)
>     return context.WithValue(ctx, append(existing, attrs...))
> }
> ```

I personally think that instead of increasing allocations with `append`, I would prefer to keep a pointer to the previous group of attributes, with something like the following (with names up for debate):

https://github.com/xorkevin/klog/blob/140a79471414b705ceefc0e9e431ed8b6650ffe6/klog.go#L271 (a toy logger I've been experimenting with that is not intended for production use yet)

```go
type (
	ctxKeyAttrs struct{}

	ctxAttrs struct {
		attrs  Attr[]
		parent *ctxAttrs
	}
)

func getCtxAttrs(ctx context.Context) *ctxAttrs {
	if ctx == nil {
		return nil
	}
	v := ctx.Value(ctxKeyAttrs{})
	if v == nil {
		return nil
	}
	return v.(*ctxAttrs)
}

func setCtxAttrs(ctx context.Context, fields *ctxAttrs) context.Context {
	return context.WithValue(ctx, ctxKeyAttrs{}, fields)
}

// CtxWithAttrs adds log attrs to context
func CtxWithAttrs(ctx context.Context, attrs ...Attr) context.Context {
	return ExtendCtx(ctx, ctx, attrs...)
}

// ExtendCtx adds log attrs to context
func ExtendCtx(dest, ctx context.Context, attrs ...Attr) context.Context {
	k := &ctxAttrs{
		parent: getCtxAttrs(ctx),
	}
	k.attrs.addAttrs(attrs)
	return setCtxAttrs(dest, k)
}
```

A log handler may then optionally choose to perform some deduplication if an attribute on a "child" context has the same key as one of its "ancestors".

Next, I agree with the concerns @jba raised here with regards to treating ctx in `args ...any` differently:

> @stephenafamo, nice idea. It does make it a little easier for people to put the context in their logs, but at the cost of adding another special case to key-value args. Also, what would we do for LogAttrs?

I'm currently of the opinion that the handler should be the only "thing" that concerns itself with what values are in a context. For example, with the above `ctxAttrs` example of mine, I would likely want to write a handler wrapping an slog.Handler which has the responsibility of reading from the context similar to the approach that @jellevandenhooff in `slogctx` takes (again names up for debate):

```go
type CtxReadingHandler struct {
	inner Handler
}

// implementation of other Handler interface methods

func (h *CtxReadingHandler) Handle(r Record) error {
	for a := getCtxAttrs(r.Context); a != nil; a = a.parent {
		r.AddAttrs(a.attrs...)
	}
	return h.inner.Handle(r)
}
```

In my view, this also has the added benefit of extending to the "opentelemetry span" use-case quite nicely, because libraries can choose to provide their own wrapping handlers such as:

```go
type OpenTelemetryLogHandler struct {
	inner slog.Handler
}

// implementation of other Handler interface methods

func (h *OpenTelemetryLogHandler) Handle(r slog.Record) error {
	if a := getOpenTelemetryLogAttrs(r.Context); a != nil {
		r.AddAttrs(a...)
	}
	return h.inner.Handle(r)
}
```

With that in mind, I think either of the following as listed by @jba in the alternatives is preferrable since it does not require the frontend `Logger` to "know" anything about the context:

> ```go
> logger.InfoCtx(ctx, "message")
> logger.Ctx(ctx).Info("message")
> ```

Am curious what people think.

**Comment by stephenafamo** [comment]
@xorkevin I had really not thought about the exact implementation.  
Maybe append is not the best use case, maybe there is a more efficient way to implement.  
I believe that we can fine-tune the exact implementation if we decide to go down this route.

>I'm currently of the opinion that the handler should be the only "thing" that concerns itself with what values are in a context.

I don't agree with this. I believe that the person logging should decide what they want to log. i.e. the logger, not the handler should decide what to include.

Also, regardless of the handler I use, it is always beneficial to be able to pass request-scoped log values.  
The openTelemetry handler could look specifically for `TraceID` and `SpanID` but what if my web server has a `RequestID` that I also want to add to each log, should I then reimplement my own handler?

>Also, what would we do for LogAttrs?

I gave a suggestion for this in [an earlier comment](https://github.com/golang/go/issues/58243#issuecomment-1421484039).

### An extra benefit

As I stated in my earlier comment, treating `context.Context` as a carrier of log attributes makes it possible to remove every other context method/field in both the logger and the handler.

Going through the main proposal issue (#56345), several people expressed concerns over having `context.Context` as a field in both the `slog.Logger` and `slog.Record` structs[^1]. Since they are used solely for passing around values, then we are ***already*** treating context as a hidden special case.

My suggestion adds 2 top level functions to the package:

1. `AddToContext`: To add attributes to a context.
2. `ContextAttrs`: To retrieve attributes from a context. Mostly useful when using `slog.LogAttrs`

And in exchange, we can remove the following:

1. `slog.Record.Context`: Replaced by resolving the attrs in any passed context instead.
3. `slog.Logger.ctx`: No longer needed since any scoped attrs can be added to the logger directly.
4. `slog.Logger.Context()`: No longer needed since the logger will not have a context.
5. `slog.Logger.WithContext()`: No longer needed since `slog.Logger.With()` does what is required.

Overall, it seems to me that it actually simplifies things. 

[^1]: See [this comment](https://github.com/golang/go/issues/56345#issuecomment-1352286205) by @jellevandenhooff

**Comment by xorkevin** [comment]
@stephenafamo thanks for your reply!

> I believe that we can fine-tune the exact implementation if we decide to go down this route.

Agreed here. Thanks for the clarification! I wanted to make sure that the actual implementation wasn't yet settled.

> it is always beneficial to be able to pass request-scoped log values

Agreed here! I don't think I made it clear enough, but I think request scoped values _should_ belong in `context.Context`. I agree with your wording that context should be treated as an "Attr carrier".

> should I then reimplement my own handler?

To me, this is something that is pretty easy to do, as demonstrated by `slogctx`. As alluded to in my previous comment, that can look like:

```go
func (h *CtxReadingHandler) Handle(r Record) error {
	for a := getCtxAttrs(r.Context); a != nil; a = a.parent {
		r.AddAttrs(a.attrs...)
	}
	return h.inner.Handle(r)
}
```

Just as you might want to implement a wrapping `net/http.ResponseWriter` or `io.Writer`, I believe that the interface for an `slog.Handler` should be (and is currently) simple enough such that anyone can easily do so.

> > Also, what would we do for LogAttrs?
>
> I gave a suggestion for this in [an earlier comment](https://github.com/golang/go/issues/58243#issuecomment-1421484039).

I unfortunately don't really agree with this:

> ```go
> 	func ContextAttrs(ctx context.Context) slog.Attr {
> 	attrs, _ := v.ctx.Value(ctxAttrs).([]slog.Attr)
> 	// I'm not sure how slog treats empty group keys. I assume they are flattened
> 	return slog.Group("", attrs...) 
> }
> ```

I guess for one, we would likely need to return `[]slog.Attr`, since according to [the docs here](https://pkg.go.dev/golang.org/x/exp/slog#Handler), an `Attr` with an empty key is ignored 

Secondly, in my mind, if we want `Attr`s from the context to be added in the arguments to a call to `Logger.LogAttr` or `Logger.Log`, then that encourages tighter coupling between third party libraries that have "request scoped variables" such as `TraceID` and `SpanID`, and `slog`, because now those libraries will either need place those attributes directly on the context as `ctxAttrs` or require people to write helper functions to do so. That's not inherently a "bad" thing, but I do not believe this is the right place for that logic.

Maybe some further clarification from @jba on what is the division of responsibility between the "frontend" `Logger` and "backend" `Handler` may be useful? To me, I view the frontend's responsibility as an entrypoint to the logging system, i.e. it should build an `slog.Record` that will be handled by the logging "backend" `slog.Handler` not unlike an `net/http.Server` which builds a `net/http.Request` to be handled by a `net/http.Handler`. Continuing the analogy, I believe that just like you may have `net/http.Handler` middleware which reads/writes values off of the request context and modifies the `net/http.Request`, an `slog.Handler` should be responsible for reading/writing values off the context, and modifying the `Attr`s on an `slog.Record`.

Based on the above, I think I am more in favor of adding `ctx context.Context` as the first argument to Logger.Log, etc. With this approach, we will still be able to remove the following:

>     2. `slog.Logger.ctx`: No longer needed since any scoped attrs can be added to the logger directly.
>
>     3. `slog.Logger.Context()`: No longer needed since the logger will not have a context.
>
>     4. `slog.Logger.WithContext()`: No longer needed since `slog.Logger.With()` does what is required.

I also just want to note that I generally am okay with increasing the api surface for "structs", but am very wary of doing the same for "interfaces" since that will make it more difficult for people to implement the interface as alluded to above. So the addition and removal of these functions is not of great concern to me. I am more concerned with the separation of responsibilities and where logic should live.

Thanks for your time and consideration!

**Comment by stephenafamo** [comment]
@xorkevin 

> To me, this is something that is pretty easy to do, as demonstrated by slogctx.  

While *technically* true, the vast majority of users will much prefer to just use one of the default handlers.  
Also, if `slog` is already propagating values with the context (as evidenced by the existence of `Record.Context`), why shouldn't it resolve those log values directly instead of many many implementations of handlers that then have to do it?

> Just as you might want to implement a wrapping net/http.ResponseWriter or io.Writer, I believe that the interface for an slog.Handler should be (and is currently) simple enough such that anyone can easily do so.

I agree with this, however, given that `slog` provides some handlers itself, and the general nature of loggers, I expect individual developers to rarely write their own handlers.

>Secondly, in my mind, if we want Attrs from the context to be added in the arguments to a call to Logger.LogAttr or Logger.Log, then that encourages tighter coupling between third party libraries that have "request scoped variables" such as `TraceID` and `SpanID`, and slog, because now those libraries will either need place those attributes directly on the context as ctxAttrs or require people to write helper functions to do so. That's not inherently a "bad" thing, but I do not believe this is the right place for that logic.

Perhaps you misunderstand how I expect this to be used. For something like open telemetry there is usually a middleware that adds request-scoped values to the context. If `slog` provides a way to add log attrs to the context, these libraries can also add them at the same time. After that, all that will be needed to add `TraceID` or `SpanID` will be adding the context to the log statement:

```go
slog.Info("a message", ctx)
```

Since `slog` will auto-extract any attrs in the context, no further change is needed for the user and they can use any existing handler.

>Based on the above, I think I am more in favor of adding ctx context.Context as the first argument to Logger.Log, etc. With this approach, we will still be able to remove the following:

This was my first suggestion too. However, it has been rejected since we don't want to make context a requirement for **every log call** even in situations where there is no context.

### Note

My suggestion is purely to make it as convenient as possible to add the context while logging. How the logger or handler resolves the attrs are not of too much concern to me.

The problem is that in the current state of `slog`, logging values from a context is quite awkward. You either need a specialized wrapper of the logger (like `slogctx`), or you need to add `WithContext(ctx)` to every output call.  
My proposal makes it extremely convenient since the developer can simply add `ctx` as one of the arguments to the output call. 

The rest of it (early resolutions, making other context methods redundant) is really just a happy coincidence although I now believe that it is more in line with how context is used in Go today.

**Comment by xorkevin** [comment]
@stephenafamo Thanks for reading, and here are my thoughts:

> Perhaps you misunderstand how I expect this to be used. For something like open telemetry there is usually a middleware that adds request-scoped values to the context. If `slog` provides a way to add log attrs to the context, these libraries can also add them at the same time. After that, all that will be needed to add `TraceID` or `SpanID` will be adding the context to the log statement:
>
> ```go
> slog.Info("a message", ctx)
> ```
>
> `slog` providing a standard way of adding log values to the context **removes** tight coupling.
>
> If each handler implementation has its own way of adding and retrieving values from the context, then how propagate log attrs through the context is directly tied to whichever handler I'm using, or I'm forced to start writing my own handler middleware.
>
> By providing a standard way, I can use `slog.AddToContext` and expect that any handler will be able to find it.

Maybe we aren't agreed on the definition of what tight coupling is. And I think I am correctly understanding what you're proposing that these libraries should do. I do not think that libraries such as the open telemetry middleware should import slog to add their request-scoped values to the context in addition to `slog.Attr`s to the context. There are a couple reasons why I currently believe this.

* First is that libraries like the open telemetry middleware now need to be aware of logging and be sure to add their request-scoped variables to the context. How will they know what is or isn't important to log? And what if you have a particular preference on what to name the keys on those attributes? This to me is an increase in coupling, i.e. one concern bleeding into another concern.
* Following from the above, assuming that now you will need to write a helper function to add the request-scope variables that you care about for logging as `slog.Attr`s to the context, now this helper will need to be called prior to every `Logger.Log` where we want those request-scoped variables. I do not think that is very elegant, nor do I think that that logic is in the right place. I would much prefer to have that logic in the Handler and modify the `slog.Record` accordingly.
* I do not think it is possible to avoid the above complexity assuming that you care about what is being logged. Either you will need to write appropriate pre-`Logger.Log` helpers to set the right request-scoped attributes on the context, or you will need to write `Handler` "middleware". I much prefer wrapping `Handler`.
* Lastly, users of these libraries may not need to implement `Handler` middleware, since library authors can do that work in supplying a default implementation, just as people have for writing middleware for `net/http.Handler`. However, I do want to note that the key advantage of this approach is that should you not like the default `Handler` "middleware" implementation, you have the option of writing your own. You aren't locked in to having particular attributes on the Record if you do not want them there.

> While _technically_ true, the vast majority of users will much prefer to just use one of the default handlers. Also, if `slog` is already propagating values with the context (as evidenced by the existence of `Record.Context`), why shouldn't it resolve those log values directly instead of many many implementations of handlers that then have to do it?

As communicated above, just wanted to reiterate that I may disagree with what attributes and/or keys for those attributes that a library author has selected for me. I want the option to use their default implementation, but also the option to write my own if I so choose.

> given that `slog` provides some handlers itself, and the general nature of loggers, I expect individual developers to rarely write their own handlers.

This I do agree with. I would like if library authors supplied their own default `Handler` "middleware" implementations.

> Since `slog` will auto-extract any attrs in the context, no further change is needed for the user and they can use any existing handler.
>
> However, it has been rejected since we don't want to make context a requirement for **every log call** even in situations where there is no context.
>
> My suggestion is purely to make it as convenient as possible to add the context while logging. How the logger or handler resolves the attrs are not of too much concern to me.
>
> The problem is that in the current state of `slog`, logging values from a context is quite awkward. You either need a specialized wrapper of the logger (like `slogctx`), or you need to add `WithContext(ctx)` to every output call. My proposal makes it extremely convenient since the developer can simply add `ctx` as one of the arguments to the output call.
>
> The rest of it (early resolutions, making other context methods redundant) is really just a happy coincidence although I now believe that it is more in line with how context is used in Go today.

Hmm, yeah I'm not too concerned what the exact api on the frontend looks like. I largely agree that I have a personal preference to make it as easy as possible to supply a context to the logger. However, I definitely am more concerned with what the responsibilites are between the frontend and backend. I did see a good suggestion on the main proposal thread https://github.com/golang/go/issues/56345#issuecomment-1423695279 that, since we do have a separate frontend and backend, maybe we can have one frontend where `context.Context` is the first argument (which may be `nil` if there is none so that we do not have to supply `context.Background()` in those situations), and another frontend where there is no context.

I guess to expound a bit why I think getting the interface and separation of responsibilites right between the frontend and backend early in the process is that those things are hard to change. Once a certain pattern is established, it becomes very sticky and hard to walk back. With a proper interface defined, I'm not too worried about what the frontend looks like exactly because that can always be easily improved over time without needing to get the entire go ecosystem to change how their `Handler`s work.

Thanks again for the discussion!

**Comment by stephenafamo** [comment]
@xorkevin 

>First is that libraries like the open telemetry middleware now need to be aware of logging and be sure to add their request-scoped variables to the context. 

By standardizing around `slog`, this is expected. I think it makes the ecosystem better that structured logging works as pretty much **plug-n-play**. 

>How will they know what is or isn't important to log? And what if you have a particular preference on what to name the keys on those attributes?

I understand this concern, a package author could make this configurable or leave it to the users to set up themselves.  
Also, similar to writing a handler middleware, a "context middleware" can also be written by hand.

>Following from the above, assuming that now you will need to write a helper function to add the request-scope variables that you care about for logging as `slog.Attrs` to the context, now this helper will need to be called prior to every `Logger.Log` where we want those request-scoped variables.

Not quite. The helper function will be run at the point where the `attrs` should be injected into the context. Every downstream `Logger.Log` call will only need to "log" the context.

>I do not think it is possible to avoid the above complexity assuming that you care about what is being logged. Either you will need to write appropriate pre-Logger.Log helpers to set the right request-scoped attributes on the context, or you will need to write Handler "middleware". I much prefer wrapping Handler.

This is true. A middleware has to exist, either in the context chain, or on the handler.

>I do want to note that the key advantage of this approach is that should you not like the default Handler "middleware" implementation, you have the option of writing your own. You aren't locked in to having particular attributes on the Record if you do not want them there.

I don't think my approach necessarily forces it either. Similar to any package author, you can write your own "context" middleware to add the attributes you desire.

>As communicated above, just wanted to reiterate that I may disagree with what attributes and/or keys for those attributes that a library author has selected for me. I want the option to use their default implementation, but also the option to write my own if I so choose.

The option remains. For example, if opentelemetry provides a handler middleware to extract `TraceID` and `SpanID` you can also write your own to extract more attributes.  
**We primarily differ in where we think the "middlewares" should be defined.**  
Reading through [this week's summary](https://github.com/golang/go/issues/56345#issuecomment-1426406720), `Attrs` are resolved as early as possible by slog itself and not the handlers. So I think my suggestion is more in the spirit of the `slog` package.

>maybe we can have one frontend where context.Context is the first argument (which may be nil if there is none so that we do not have to supply context.Background() in those situations), and another frontend where there is no context.

This is always possible, just as how people can (and may continue to) use different loggers entirely.  
However, I believe that `slog` should account for such a common use-case. Personally, if no further changes were made to the package, I would likely use something like `slogctx` instead and would therefore prefer if the base package accounted for this. 

Finally, I want to restate that I primarily want a mechanism to be able to write `slog.Info("a message", ctx)`. If it is later decided that the extraction should happen in the Handler vs the Logger, I really don't mind too much.

---

* Reading through [last week's summary](https://github.com/golang/go/issues/56345#issuecomment-1413742844), groups with empty keys are inlined so my rough implementation would work 😁
* Treating `context.Context` specially, may make it easy to "enforce" context logging through linters by checking that every log line in a function that accepts a context logs the context.

**Comment by xorkevin** [comment]
@stephenafamo

> By standardizing around `slog`, this is expected. I think it makes the ecosystem better that structured logging works as pretty much **plug-n-play**.
>
> > How will they know what is or isn't important to log? And what if you have a particular preference on what to name the keys on those attributes?
>
> I understand this concern, a package author could make this configurable or leave it to the users to set up themselves. Also, similar to writing a handler middleware, a "context middleware" can also be written by hand.

I think you probably don't disagree with this, but I just want to note that regardless of whether we go with "frontend" middleware, or `Handler` "middleware", I would strongly prefer that "canonical" attribute keys added by the proposed `AddToContext` should not be added to the context by library authors automatically in areas such as this open telemetry middleware that we've mentioned in our discussions. Regardless of whether the attribute is read pre-`Logger.Log` or in `Handler.Handle`, we want to give library users the option to not use the default slog middleware and customize what gets logged in their own "frontend" or `Handler` middleware. To do otherwise would be tight coupling.

> Not quite. The helper function will be run at the point where the `attrs` should be injected into the context. Every downstream `Logger.Log` call will only need to "log" the context.

Agreed, I think I worded this poorly. The helper function would need to be run at some point prior to a `Logger.Log` that wants to use it.

> The option remains. For example, if opentelemetry provides a handler middleware to extract `TraceID` and `SpanID` you can also write your own to extract more attributes. **We primarily differ in where we think the "middlewares" should be defined.** Reading through [this week's summary](https://github.com/golang/go/issues/56345#issuecomment-1426406720), `Attrs` are resolved as early as possible by slog itself and not the handlers. So I think my suggestion is more in the spirit of the `slog` package.

I wanted to note that the summary is not referring to the frontend. It is referring to the `slog` package. I do not think that there is any additional evidence for or against a "frontend" or `Handler` "middleware" being more in the spirit of the `slog` package than last week. The exact change that was made, as noted in the documentation is the following:

https://pkg.go.dev/golang.org/x/exp/slog#Record.AddAttrs: AddAttrs appends the given Attrs to the Record's list of Attrs. It resolves the Attrs before doing so.

> This is always possible, just as how people can (and may continue to) use different loggers entirely. However, I believe that `slog` should account for such a common use-case. Personally, if no further changes were made to the package, I would likely use something like `slogctx` instead and would therefore prefer if the base package accounted for this.

I think we agree here? As in, exactly how a ctx may be supplied doesn't really matter because there will likely always be other frontend implementations. Agreed that `slog` should try and handle the common simple use-case, of which supplying a ctx is one of them.

> Finally, I want to restate that I primarily want a mechanism to be able to write `slog.Info("a message", ctx)`. If it is later decided that the extraction should happen in the Handler vs the Logger, I really don't mind too much.

I feel the need to restate that I personally like a lot of your proposed suggestions. And I do not view our viewpoints as mutually exclusive. Exactly where the ctx is provided isn't the greatest concern to me, because, again, if people are unhappy, a different "frontend" is probably pretty easy to create with the exact arguments present that are applicable for anyone's particular preference. The one thing that I strongly disagree with is the early resolution of attributes by the "frontend" `Logger`. It seems like you don't have too strong of an opinion here which is maybe a good thing? I strongly believe that the `Handler` should be the thing that is responsible for adding attributes to the record for all the aforementioned reasons:

1. separation of concerns: Again, quoting the proposal:

    > having a flexible backend provides several benefits: programs can display the logs in a variety of formats, convert them to an RPC message for a network logging service, store them for later processing, and add to or modify the data.

    I think reading attributes from the context and adding them to the Record should be the `Handler`'s responsibility.

2. In my view, `Handler` middleware is an easier way to customize what does and doesn't get logged. In the "frontend middleware" case, you would have to update the helper function everywhere in the codebase that uses it. Which may or may not be numerous. With `Handler` middleware, that only happens in one place.

Am curious what you think! Thanks!

**Comment by stephenafamo** [comment]
I think we are mostly in agreement @xorkevin 

Since I am primarily concerned about the ease of logging contexts, I can mostly think of a way to do this by resolving context attributes early.

Where the middlewares have to be placed to extract the context values, I have no strong opinions about.

**Comment by jba** [comment]
@stephenafamo and @xorkevin, thanks for your detailed exchange. It's given me a lot to think about.

What if slog made it easy to pass the context itself as an attribute, say by defining
```
const ContextKey string = "ctx"
```
and maybe for convenience
```
func Context(ctx context.Context) Attr { return Any(ContextKey, ctx) }
```
slog itself wouldn't do anything with the context, but a handler could extract whatever it wanted from it, or just ignore it.

We could then get rid of the special-purpose context machinery.

The next step would treat a bare `context.Context` in key position as a shorthand for `Context(ctx)`, so you could write
```
log.Info("msg", ctx)
```
In my opinion, that would definitively solve the problem I posed in this issue, which is how to make it easy to add a context to log statements. I'm not sure we should take that step, but at least it is a relatively small one. Even if we don't, anyone could define
```
var C = slog.Context
```
and write `log.Info("msg", C(ctx))`.

If I've understood the discussion, I've adopted the syntax ideas of @stephenafamo but follow @xorkevin in preferring that handlers extract attributes from the context.

**Comment by stephenafamo** [comment]
@jba If we are adding the context as an attr with attrkey, then by default the context will sort of be just logged however `fmt` stringifies the context value correct?

If the intention is that a handler middleware can extract the values it needs and then drop the context from the attributes, it then becomes an issue to chain these middlewares because an earlier middleware could drop the context before the necessary values are extracted.

A possible solution may be to skip logging attributes with a context key in the default `slog` handlers.

**Comment by jba** [comment]
> If we are adding the context as an attr with attrkey, then by default the context will sort of be just logged however fmt stringifies the context value correct? ... A possible solution may be to skip logging attributes with a context key in the default slog handlers.

@stephenafamo, exactly: the built-in handlers would ignore an Attr whose key equaled ContextKey.

> an earlier middleware could drop the context before the necessary values are extracted.

I'd suggest that middleware handlers don't drop the context, just add attrs. That's faster anyway. The output handler can easily drop the context in its final traversal of the attrs.

**Comment by jba** [comment]
> Next would be to create a linter (or add a check to go vet) to ensure that logs inside a function that accepts a context, includes the context.

You could write that linter with the current design too (look for calls to `WithContext`), but it would have a lot of false positives for functions that don't create the logger themselves, because it wouldn't know if the logger obtained by the function already had the context in it. I think an accurate linter would be hard to write (would require interprocedural control-flow analysis).

**Comment by stephenafamo** [comment]
@jba This seems to be a nice solution. However, I wonder if this will lead to performance issues. 

From what I understand, to extract one-or-more values from the context, a handler middleware will have to loop through **every** attribute of **every** record, check if the key matches the context key, and check if the context contains the key(s) they are interested in. This already seems like a lot of work to do on **every** log record, and considering that adding multiple middlewares may be encouraged with this design, it could get even worse.

Or does the design prevent this from being an issue?

**Comment by jba** [comment]
@stephenafamo, you're right about the potential performance impact. It's going to be similar for many middleware handlers, like one that removes PII or changes key names. I think people who care that much about performance will avoid middleware, or design their Handler to work in a single pass, maybe with something like a ReplaceAttr function.

**Comment by xorkevin** [comment]
@jba 

> We could then get rid of the special-purpose context machinery.

I'm not sure if we want to entirely do this, due to what @stephenafamo wrote here:

> From what I understand, to extract one-or-more values from the context, a handler middleware will have to loop through every attribute of every record

I think the `slogctx` package itself is a good example of how I would want to use a `Handler`.

I think keeping `Record.Context` makes sense to make it easy for `Handler`s to interact with the context. This would also likely have the `slog.Logger` frontend loop through the attributes being added and pick out the last one with a key of `ContextKey` if it exists.

**Comment by stephenafamo** [comment]
@xorkevin I don't think any changes need to be made to the front end logger aside from treating a bare `ctx` as `slog.Context(ctx)`

Slog allows duplicate keys so there is no need to check for the last one.

**Comment by xorkevin** [comment]
@stephenafamo 

> Slog allows duplicate keys so there is no need to check for the last one.

Gotcha. I meant this insofar as setting a `Record.Context` which I think we should keep. Whether or not we add the `ctx` as an `Attr` and have the final output `Handler` ignore it I think I'm okay with either. `Record.Context` I think is necessary to allow `Handler`s to more efficiently interact with the context.

**Comment by stephenafamo** [comment]
@xorkevin 

> Gotcha. I meant this insofar as setting a `Record.Context` which I think we should keep. Whether or not we add the `ctx` as an `Attr` and have the final output `Handler` ignore it I think I'm okay with either. `Record.Context` I think is necessary to allow `Handler`s to more efficiently interact with the context.

I am not a fan of contexts as struct fields, so I'll be happy to see `Record.Context` go.

However, the bigger issue is that having multiple ways to pass a context makes things messier and Go has the tendency to have one way to do things. For example, in this scenario, Handler implementation will now have to check 2 places for context values.

**Comment by xorkevin** [comment]
@stephenafamo 

> I am not a fan of contexts as struct fields, so I'll be happy to see `Record.Context` go.

I generally agree, though in this case the alternative is having context in the struct `Attr`, of which there could also potentially be multiple. To me `Record.Context` feels okay in this case because `Record` is essentially a bunch of "arguments" that are being passed to the `Handler`, similar to how `ctx` should ordinarily be passed as an argument. Most ideal of course would be to pass `ctx` explicitly through a `LogCtxAttrs(ctx, attrs...)`, through to `Handler(ctx, Record)`.

**Comment by seankhliao** [comment]
Today having converted a codebase from an internal library that had a similar api as `slogctx` to `slog`, the current `WithContext` felt quite unnatural to type and easy to leave off as every other attribute was part of args to `Info()`.

```go
logger.WithContext(ctx).Info("msg", myKey, myValue, anotherKey, anotherValue)
```

I think having `func Context(ctx context.Context) Attr` would be an improvement, placing it with other attributes, (bikeshed: shorten to Ctx? ref #42877):

```go
logger.Info("msg", slog.Context(ctx), myKey, myValue, anotherKey, anotherValue)
logger.Info("msg", slog.Ctx(ctx), myKey, myValue, anotherKey, anotherValue)
```

Adding another special case for how args is processed though feels troubling (bare `ctx` arg), the function signature for `slog.Info` is already opaque enough. I'm suggesting https://github.com/golang/go/issues/56345#issuecomment-1428448297 to separate out the alternating kv pairs, which would put `slog.Context` on equal footing in terms of weight of arguments while making it clearer what the functions accept.

```go
logger.Info("msg", slog.Context(ctx), slog.KV(myKey, myValue, anotherKey, anotherValue))
logger.Info("msg", slog.Context(ctx), slog.String(myKey, myValue), slog.String(anotherKey, anotherValue))
```

**Comment by jba** [comment]
>  I think is necessary to allow Handlers to more efficiently interact with the context.

Somewhere, the final Handler is going to do
```
r.Attrs(func(a slog.Attr) {...})
```
That callback will already have a test for groups (since they will almost certainly be processed recursively). Another check for context won't slow things down much.

Yes, middleware handlers would be slower. But people who care about speed will probably try to avoid middleware anyway and collapse all the logic to one handler.

To be clear, I'm not strongly in favor of making context an attr;  I could go either way. Just playing out the arguments.

**Comment by jba** [comment]
> It seems like just keeping it simple and allowing a LogCtx(ctx, ...) and LogAttrsCtx(ctx, ...) would simplify a lot of code in exchange for "just one more" argument.

@xorkevin, you couldn't stop there, you'd need InfoCtx, DebugCtx, WarnCtx and ErrorCtx. And now you've added six methods. Then you'd need six more functions for the default logger.

**Comment by seankhliao** [comment]
Isn't context already special in that it has to be kept around for a call the `handler.Enabled(ctx, lvl)`?

**Comment by xorkevin** [comment]
@jba 

> Yes, middleware handlers would be slower. But people who care about speed will probably try to avoid middleware anyway and collapse all the logic to one handler.

I think we are agreed here. If there is no middleware, then there is no difference. I'm not sure if this was your original intention behind the proposal, but I think the beauty of `Handler` as an interface is to allow other "wrapping" implementations, just like one might with `http.ResponseWriter`, or `io.Writer`. That personally feels very "go"-like in my experience, but I understand if that goes against the spirit of the proposal.

> To be clear, I'm not strongly in favor of making context an attr; I could go either way. Just playing out the arguments.

Am happy to chat more about this! After reading everyone's thoughts, I'm currently still pretty strongly of the opinion that we want to keep `Record.Context` or passing `ctx` directly to `Handler.Handle(ctx, ...)`.

> you'd need InfoCtx, DebugCtx, WarnCtx and ErrorCtx. And now you've added six methods. Then you'd need six more functions for the default logger.

Definitely agreed here, though I personally don't mind too much. I think I'd be able to get by with `Logger.LogAttrCtx` and writing the appropriate helpers around that.

**Comment by AndrewHarrisSPU** [comment]
> To be clear, I'm not strongly in favor of making context an attr; I could go either way. Just playing out the arguments.

For the sake of argument ... my sense is that decisions about logging and contexts are made at package-level granularity. There are a few categories:

1. A package that wants to both instrument and observe contexts for logging imports both `slog` and a third party `telemetry`. I think this is ideal as far as reasoning locally about what is happening in the package, or inferring what the package author wants to accomplish.
2. A package that just observes contexts for logging doesn't import `telemetry`, but does import `slog` may reliably observe context in `Record`-producing calls. This is probably necessary if code wants to decouple from any particular `telemetry`, and fairly obvious if the whole application is highly saturated with `context`.
3. A package that doesn't import `telemetry`, but does import `slog` may also _*not*_ reliably observe context in `Record`-producing calls.

I think it's plausible that a package author may want to assert that they are in category (2), and also plausible for an application developer to do something clever around category (3) to get consistent observation. Context-as-attribute makes category (2) a bit more munge, a bit less static in nature - my impression is I would rather isolate this decision once in an import statement or some other package-level mechanism, than have it inflect every `Record`-producing call site. Also, I think it's implausible to ask or expect packages in category (3) to add contexts with context-as-attribute, and the context-in-logger option might be preferred here. I'm enthusiastically for having a single, common symbol for storing attributes on a context, less so about passing context-as-attribute.

**Comment by jba** [comment]
> Context-as-attribute makes category (2) a bit more munge, a bit less static in nature - my impression is I would rather isolate this decision once in an import statement or some other package-level mechanism, than have it inflect every Record-producing call site. 

@AndrewHarrisSPU, can you explain that in more detail? Consider an exported function in a package:
```
package p

func F(ctx context.Context, ...) {...}
```
If `F` wants to log (via a Logger passed as an argument, say), what package-level code could get its context to the Handler without writing something at the call site?

> Also, I think it's implausible to ask or expect packages in category (3) to add contexts with context-as-attribute, and the context-in-logger option might be preferred here.

I don't see the distinction here, because I can always write
```
logger = logger.With(slog.Context(ctx))
```
to get a similar effect to
```
logger = logger.WIthContext(ctx)
```

**Comment by xorkevin** [comment]
@jba 

Just wanted to get your thoughts on @seankhliao 's comment here:

> Isn't context already special in that it has to be kept around for a call the handler.Enabled(ctx, lvl)?

To me this means that we would likely also want to obtain the context in the body of `Logger.Log*`. In my opinion, it would make sense then to pass that context that we would obtain as the first argument to `Handler.Handle(ctx, ...)` or keeping `Record.Context`. Working backwards from there, I think it could also make sense to allow people to pass a `nil`-able `ctx` as the first argument to all `Log`, `Info`, etc. `Logger` methods.

**Comment by AndrewHarrisSPU** [comment]
> @AndrewHarrisSPU, can you explain that in more detail?

Sure, and let me quickly remark that I don't think context-as-attribute would be a broken feature, but I just want to highlight where I'm not convinced it's covering some cases. This is a really long discussion and I appreciate the effort at wrangling these details. I really would love to see `slog` in the standard library and think it's complete, it's still great to really focus on contexts in detail to see if there's anything more that would help.

Here's a scenario from the category where we could suppose the author of `package p` clearly understands that they need contexts for each logging call. The struct `p.S` is relatively long-lived, more than the `context`s or `span`s that should be logged in method calls like `p.S.M`. Then, it's not well motivated to construct an instance of `p.S` with a `p.S.log.ctx` field set. But it's still true (and maybe worth statically checking) that each logging call in methods of `p.S` require a context.

```go
package main

func main() {
    log := slogctx.New(slog.New(...))
    s := p.NewS(log)

    ctx := telemetry.Start(context.Background(), "main")

    go func() {
        s.M(ctx)
    }()
    ...
}
```

```go
package p

type S struct {
    id string
    log *slogctx.Logger
}

func NewS(log *slogctx.Logger) *S {
    s := new(S)
    ...
    s.log = log.With("id", s.id)
    ...
    return s
}

func (s *S) M(ctx context.Context) {
    s.log(ctx, "I'm in M on an S")
}
```

I think the logger-embedded-in-a-struct solution highlights that there is a place for a `Logger` that is always going to want to consume a context, but that context is not one that should be set on the `Logger.ctx` field. It aligns with my priors from adventures in logging, I have in the past preferred statically checked front-end discipline here. In this scenario I don't think context-as-attribute prevents me from doing anything, it's just that it doesn't statically check that all calls include a context.

---

On the other category of working out how to put match context-aware logging with packages or API that isn't context aware - I agree there's not much difference between `logger = logger.With(slog.Context(ctx))`, `logger = logger.WIthContext(ctx)`. 

To be sharper - there was discussion of removing the `Logger.ctx` field, both of these seem to require the `Logger.ctx` field. Maybe context-as-attribute and context-in-Logger aren't ultimately exclusive, but `Logger.ctx`, and context-in-`Record` seem like they would be essential here when callee/caller are mismatched a bit on conventions.

**Comment by stephenafamo** [comment]
@AndrewHarrisSPU I struggle to think of a scenario where a static context on the logger will be useful beyond setting static attributes. Please help me understand 🙏🏾 

Also, a similar effect can be achieved by doing:

```go
var logger slog.Logger // an existing logger we want to add a static context to
logger = logger.With(slog.Context(ctx)) // added like any other static attribute to the logger
```

@jba I just noticed that `logger.With()` takes `...any`. While I understand making usage less verbose by having `Debug|Warn|Info|Error` take key-value pairs, does `logger.With` have to be the same? I think it would be better to change it  to `logger.With(...Attr)`

**Comment by AndrewHarrisSPU** [comment]
> @AndrewHarrisSPU I struggle to think of a scenario where a static context on the logger will be useful beyond setting static attributes. Please help me understand 🙏🏾

I am reviewing `LogValuer`-related changes in the `slog` repo to check my assumptions, as the details are in flux and subtle, but ... a subtle implementation detail that I find very interesting is that `Record`s resolve `LogValuer`s when they are added with `Record.AddAttrs`.

An open question, if adding a `LogValuer` to a context, when is it resolved? I think at `Record` time or sort of immediately after is what I might anticipate, at least. If this is the case, one succinct way to define something that gets freshly evaluated each time it's expanded in a `Record`:

```go
type LogThunk func() slog.Value

func (fn LogThunk) LogValue() slog.Value {
    return fn()
}
```

One result of pulling a trick like this is, if I store the context instrumented with the `LogValuer` in the `Logger`, and pass the `Logger` to context-unaware API, I still get a dynamic evaluation of the `LogValuer` each time the context-unaware API logs. It's not ideal, but it's not nothing.

**Comment by jba** [comment]
> While I understand making usage less verbose by having Debug|Warn|Info|Error take key-value pairs, does logger.With have to be the same? I think it would be better to change it to logger.With(...Attr)

@stephenafamo, our view is that  `...any` is the norm; this is the everyday form that most people will use. `LogAttrs` is an optimization. Optimizing Logger.With isn't as important because we don't expect it to be called as frequently as a log output method.

**Comment by stephenafamo** [comment]
> @stephenafamo, our view is that `...any` is the norm; this is the everyday form that most people will use. `LogAttrs` is an optimization. Optimizing Logger.With isn't as important because we don't expect it to be called as frequently as a log output method.

@jba I thought that `...any` while the norm, is for convenience.

I am not suggesting `...Attr` for optimization, but rather for correctness. As you've said, `Logger.With` will not be called frequently, so wouldn't it be better to strive for correctness?  

- No need for `"!BADKEY"`  
- No need for treating special cases  
- No need for special formatting for readability  

---

Perhaps it is better to have this discussion on the original issue instead of here 🤔

**Comment by jba** [comment]
>  Records resolve LogValuers when they are added with Record.AddAttrs.

@AndrewHarrisSPU, that is a recent change, to make invocations of `LogValue` more predictable.

> An open question, if adding a LogValuer to a context, when is it resolved? 

Right now, never. Unless the Handler does it. Are you assuming some change to slog?

> if I store the context instrumented with the LogValuer in the Logger, and pass the Logger to context-unaware API, I still get a dynamic evaluation of the LogValuer each time the context-unaware API logs.

Nothing in slog right now will pull that LogValuer out of the context. Can you elaborate on how this works?

**Comment by jba** [comment]
> Perhaps it is better to have this discussion on the original issue instead of here

Yes, it would be good for posterity if this suggestion was on the proposal issue. However, we've been down this road before and I don't see anything in your argument that would change my view. I don't think any of the negatives you mention are serious problems in practice.

**Comment by stephenafamo** [comment]
> Yes, it would be good for posterity if this suggestion was on the proposal issue. However, we've been down this road before and I don't see anything in your argument that would change my view. I don't think any of the negatives you mention are serious problems in practice.

In that case there's no need

**Comment by AndrewHarrisSPU** [comment]
@jba
> > An open question, if adding a LogValuer to a context, when is it resolved?
> 
> Right now, never. Unless the Handler does it. Are you assuming some change to slog?
> 
> > if I store the context instrumented with the LogValuer in the Logger, and pass the Logger to context-unaware API, I still get a dynamic evaluation of the LogValuer each time the context-unaware API logs.
> 
> Nothing in slog right now will pull that LogValuer out of the context. Can you elaborate on how this works?

I was thinking of some change, along the lines of adding the common/open store for attributes @stephenafamo suggested.

Just a quick PoC to demonstrate (confirm to myself, even...) that I can store a `LogValuer` that is evaluated per-`Record`: https://go.dev/play/p/P9ScAjzpiR6

In terms of slogging configuration, this defines a `Handler` that expands attributes found on a context per-`Record`, and passes a bigger `Record` to a `JSONHandler`. The per-instance state that is added to the the `Handler` chain is, in this PoC, a list of context keys to observe.

Relative to this PoC, I think the common/open store (one context key or lookup function that `Handler` authors could expect to use, and pull a number of attributes from) is a great idea - it would be a quality of life improvement and maybe a bit more performant. It would remove the need for state related to context keys on `Handler`s that just wanted to observe an open store.

**Comment by jba** [comment]
This discussion has been extremely helpful. See [my comment on the proposal issue](https://github.com/golang/go/issues/56345#issuecomment-1435247981 ) for how I think we should proceed.

**Comment by suiriass** [comment]
Why the LogDepth function is not exported in the new version Does custom calldepth not work properly @jba

**Comment by jba** [comment]
@suiriass, because now you can write it yourself: https://go.googlesource.com/exp/+/refs/heads/master/slog/example_depth_test.go#19.

**Comment by lictw** [comment]
Hello! I can't handle such huge discussion (not native speaker), but I see references on `slog.Context(ctx)` method as an alternative, what is it and why I don't see that in 1.21 slog package? Can release slog version log context values out of box?

**Comment by jba** [comment]
@lictw If you want your handler to get values out of a context, you will have to call it with `slog.InfoContext` instead of `slog.Info`, and similarly for `Debug`, `Warn` and `Error`. We got rid of `slog.Context` a while ago.

**Comment by lictw** [comment]
`slog.Context` was for the same, so just to pass ctx to handler? Or was it method to transform context values into Attrs?

**Comment by jba** [comment]
Yes, it was a way to pass a context to handlers without defining a bunch of methods that took a context (which is what we eventually did anyway).

There is nothing in slog to transform context values into Attrs. I don't think there is a single way to do that which stands out as the best.

**Comment by lictw** [comment]
But, @jba, if you delete and `slog.Context`, and `handler.WithContext` in final, now if I want to pass logger into function and I want that logger has the same context (just to logging, context hold something like session_id) I should or pack logger inside context (and it's the only option to pass via single argument), or pass as two arguments, right? Or have a function like `ctxToAttrs() []slog.Attrs` and pass single logger using `logger.With()`, but the first: it's what my handler do (extracting), the second: for what now I have `logger.*Context` methods, I could use `logger.Info("", ctxToAttrs(...))`, so, miss of `handler.WithContext` looks very strange..

**Comment by jba** [comment]
I think it's fine to pass the logger in the context. (Many people disagreed, which is why we left it out.)

If there are specific things you want to get from the context, I think you should write a handler to get them, and use the logger.*Context methods. That is easier than calling ctxToAttrs.

**Comment by lictw** [comment]
> If there are specific things you want to get from the context, I think you should write a handler to get them, and use the logger.*Context methods. That is easier than calling ctxToAttrs.

**Comment by lictw** [comment]
> If there are specific things you want to get from the context, I think you should write a handler to get them, and use the logger.*Context methods. That is easier than calling ctxToAttrs.
So pass and logger, and context into function is fine for you too? (if not to pass logger with context) I see redundancy in this, it's a pity there is no option how to pass a logger/handler with context, so even if child function will just use *WithoutContext* methods parent context will be available for handler.

**Comment by muhlemmer** [comment]
Now that I've written my own handler that takes data from the context (most importantly a tracing id) and pass it on to a wrapped logger. There is one pitfall:

 If `logger.WithGroup()` is called and the new logger is used with the context, the attributes end up in that group. At the moment of logging there is no obvious way to get back to the parent group / handler. Perhaps I'm missing something:

```go
package main

import (
	"context"
	"log/slog"
	"os"
)

type idHandler struct {
	slog.Handler
}

func (h *idHandler) Handle(ctx context.Context, record slog.Record) error {
	id, ok := ctx.Value(idKey).(int)
	if ok {
		record.AddAttrs(slog.Int("id", id))
	}
	return h.Handler.Handle(ctx, record)
}

func (h *idHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
	return &idHandler{
		Handler: h.Handler.WithAttrs(attrs),
	}
}

func (h *idHandler) WithGroup(name string) slog.Handler {
	return &idHandler{
		Handler: h.Handler.WithGroup(name),
	}
}

type idKeyType struct{}

var idKey idKeyType

func main() {
	logger := slog.New(&idHandler{slog.NewTextHandler(os.Stdout, nil)})

	ctx := context.WithValue(context.TODO(), idKey, 42)

	logger = logger.With("Hello", "World")
	logger.InfoContext(ctx, "lets log!")

	logger = logger.WithGroup("group").With("foo", "bar")
	logger.InfoContext(ctx, "lets log!")
}
```

Output:

```
time=2023-08-24T22:34:36.481+03:00 level=INFO msg="lets log!" Hello=World id=42
time=2023-08-24T22:34:36.481+03:00 level=INFO msg="lets log!" Hello=World group.foo=bar group.id=42
```

`id` now became part of the group. Same grouping happens with the JSON handler. I don't think that would make it easy to parse and group these logs based on `id`.

**Comment by jba** [comment]
@muhlemmer that's unfortunate, but that is a consequence of the design. The final handler, or the processor consuming the logs, will have to recognize that attribute specially.

**Comment by yysushi** [comment]
@muhlemmer @jba  

how about this code?

<details>

```go
package main

import (
	"context"
	"log/slog"
	"os"
)

type idHandler struct {
	slog.Handler
	group string
}

func (h *idHandler) Handle(ctx context.Context, record slog.Record) error {
	id, ok := ctx.Value(idKey).(int)
	if ok && h.group == "" {
		record.AddAttrs(slog.Int("id", id))
	}
	return h.Handler.Handle(ctx, record)
}

func (h *idHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
	if h.group == "" {
		return &idHandler{
			Handler: h.Handler.WithAttrs(attrs),
		}
	}
	var groupedAttrs []slog.Attr
	for _, attr := range attrs {
		groupedAttrs = append(groupedAttrs, slog.Group(h.group, attr))
	}
	return &idHandler{
		Handler: h.Handler.WithAttrs(groupedAttrs),
		group:   h.group,
	}
}

func (h *idHandler) WithGroup(name string) slog.Handler {
	return &idHandler{
		Handler: h,
		group:   name,
	}
}

type idKeyType struct{}

var idKey idKeyType

func main() {
	logger := slog.New(&idHandler{Handler: slog.NewTextHandler(os.Stdout, nil)})

	ctx := context.WithValue(context.TODO(), idKey, 42)

	logger = logger.With("Hello", "World")
	logger.InfoContext(ctx, "lets log!")

	logger = logger.WithGroup("group").With("foo", "bar")
	logger.InfoContext(ctx, "lets log!")
}
```
</details>

**Comment by jba** [comment]
@yysushi You don't apply the groups from `WithGroup` to the attrs in the `Record`. But if you did that and then added the "id" attribute, then I think that would work as long as `WithGroup` is never called on the handler at the end of the chain.
